### PR TITLE
add support for multiple configurations for debugger

### PR DESCRIPTION
**Pull request by shatzi**
# What Does This Do

Adding support for multiple configuration for debugger. This allow backend to send multiple probe configuration to be applied at the client library

# Motivation

Support for logs probes.

# Additional Notes

### CODE REVIEW

File: remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java [hunk 1]
```diff
@@ -321,249 +342,64 @@ private void handleAgentResponse(ResponseBody body) {
...
+              "Told to handle config key "
+                  + configKey
+                  + ", but the product "
+                  + parsedConfigKey.getProductName()
+                  + " is not being handled");
+        }
+        if (!(sortedKeys.containsKey(product))) {
+          sortedKeys.put(product, new ArrayList<>());
+        }
+        sortedKeys.get(product).add(parsedConfigKey);
```
Comment (jpbempel) [inline remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java:364]: double lookup
use `computeIfAbsent` method instead

File: remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  private static final Pattern EXTRACT_PRODUCT_REGEX =
+      Pattern.compile("([^/]+)(/\\d+)?/([^/]+)/([^/]+)/config");
+  private final String org;
+  private final Optional<Integer> version;
+  private final String productName;
+  private final String configId;
+  private final Product product;
+
+  ParsedConfigKey(String org, Optional<Integer> version, String productName, String configId) {
```
Comment (jpbempel) [inline remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java:20]: Optional sounds useless here, could be just an Integer.

File: remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+  byte[] getTargetFileContent(RemoteConfigResponse fleetResponse, String configKey) {
+    // fetch the content
+    Optional<byte[]> maybeFileContent;
+    try {
+      maybeFileContent = fleetResponse.getFileContents(configKey);
+    } catch (Exception e) {
+      throw new ConfigurationPoller.ReportableException(e.getMessage());
+    }
+    if (!maybeFileContent.isPresent()) {
```
Comment (jpbempel) [inline remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java:151]: Optional does not bring any value is we just check isPresent. equivalent to check against null here.

File: remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java (hunk 1 above)
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java:364]: nice

File: remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java (hunk 2 above)
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java:20]: changed

File: remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java (hunk 3 above)
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java:151]: removed

File: remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+      } catch (ConfigurationPoller.ReportableException e) {
+        this.error = e;
+      }
+    }
+
+    for (String configKey : cachedTargetFiles.keySet()) {
+      if (!configBeenUsedByProduct.contains(configKey)) {
+        changesDetected = true;
+        callListenerRemoveTarget(hinter, listener, ParsedConfigKey.parse(configKey));
```
Comment (jpbempel) [inline remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java:70]: The back and forth on parsed and string representation of ConfigKey is painful, could we harmonize that?

File: remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java [hunk 5]
```diff
@@ -321,249 +342,61 @@ private void handleAgentResponse(ResponseBody body) {
...
-        this.ratelimitedLogger.warn(
-            "Error processing config key {}: {}", configKey, e.getMessage(), e);
-        this.collectedCfgErrors.put(configKey, e.getMessage());
-        failures++;
-      }
-    }
+    ReportableException error = null;
 
-    updateNextState(fleetResponse, inspectedConfigurationKeys, errorMessage);
+    Map<Product, List<ParsedConfigKey>> sortedKeys = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java]: renamed

File: remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java [hunk 6]
```diff
@@ -65,16 +67,14 @@
   private final long maxPayloadSize;
   private final boolean integrityChecks;
 
-  /** Map from product name to deserializer/listener. */
-  private final Map<String, DeserializerAndListener<?>> listeners = new HashMap<>();
+  private final Map<Product, ProductState> productState = new HashMap<>();
+  private final Map<Product, ProductListener> productListeners = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java]: merged

File: remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+          + this.version
+          + "/"
+          + this.productName
+          + "/"
+          + this.configId
+          + "/config";
+    } else {
+      return this.org + "/" + this.productName + "/" + this.configId + "/config";
+    }
+  }
```
[the remark this answers is not in the document]
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/state/ParsedConfigKey.java]: kept originalKey to be used as equals, hash and toString

File: remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java (hunk 4 above)
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/state/ProductState.java]: trying to improve this as much as I can

**Review by jpbempel — APPROVED** [review]

File: remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java [hunk 8]
```diff
@@ -321,249 +339,61 @@ private void handleAgentResponse(ResponseBody body) {
...
+          throw new ReportableException(
+              "Told to handle config key "
+                  + configKey
+                  + ", but the product "
+                  + parsedConfigKey.getProductName()
+                  + " is not being handled");
+        }
+        parsedKeysByProduct.computeIfAbsent(product, k -> new ArrayList<>()).add(parsedConfigKey);
+      } catch (ReportableException e) {
+        error = e;
```
[the remark this answers is not in the document]
Comment (shatzi) [reply remote-config/src/main/java/datadog/remoteconfig/ConfigurationPoller.java:360]: yeah. we only have one error to report on the message. but I think we can do better. let me fix that

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerProductChangesListener.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    if (configKey.getConfigId().startsWith("snapshotProbe_")) {
+      SnapshotProbe snapshotProbe = Adapter.deserializeSnapshotProbe(content);
+      configChunks.put(configKey.getConfigId(), (builder) -> builder.add(snapshotProbe));
+    } else if (configKey.getConfigId().startsWith("metricProbe_")) {
+      MetricProbe metricProbe = Adapter.deserializeMetricProbe(content);
+      configChunks.put(configKey.getConfigId(), (builder) -> builder.add(metricProbe));
+    } else if (configKey.getConfigId().startsWith("logProbe_")) {
+      LogProbe logProbe = Adapter.deserializeLogProbe(content);
+      configChunks.put(configKey.getConfigId(), (builder) -> builder.add(logProbe));
+    } else {
```
[the remark this answers is not in the document]
Comment (shatzi) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/DebuggerProductChangesListener.java:85]: added verification for guid config id

**Review by ojung — APPROVED** [review]

**Review by smola — APPROVED** [review]

**Pull request merged**